File: cpp/matrix.cpp
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <iomanip>

// Matrix multiplication implementation
std::vector<std::vector<double>> matrix_multiply(
    const std::vector<std::vector<double>>& matrix_a,
    const std::vector<std::vector<double>>& matrix_b) {
    
    size_t rows_a = matrix_a.size();
    size_t cols_a = matrix_a[0].size();
    size_t cols_b = matrix_b[0].size();

    // Initialize result matrix with zeros
    std::vector<std::vector<double>> result(rows_a, std::vector<double>(cols_b, 0.0));

    // Perform multiplication
    for (size_t i = 0; i < rows_a; i++) {
        for (size_t j = 0; j < cols_b; j++) {
            for (size_t k = 0; k < cols_a; k++) {
                result[i][j] += matrix_a[i][k] * matrix_b[k][j];
            }
        }
    }

    return result;
}

// Run matrix multiplication benchmark with 100x100 matrices
void run_benchmark() {
    const int size = 100;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);

    // Generate two random 100x100 matrices
    std::vector<std::vector<double>> matrix_a(size, std::vector<double>(size));
    std::vector<std::vector<double>> matrix_b(size, std::vector<double>(size));
    
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            matrix_a[i][j] = dis(gen);
            matrix_b[i][j] = dis(gen);
        }
    }

    // Measure multiplication time
    auto start_time = std::chrono::high_resolution_clock::now();
    auto result = matrix_multiply(matrix_a, matrix_b);
    auto end_time = std::chrono::high_resolution_clock::now();
    auto execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

    std::cout << "Test: Matrix Multiplication" << std::endl;
    std::cout << "Matrix size: " << size << "x" << size << std::endl;
    std::cout << "Execution time: " << execution_time << " ms" << std::endl;
    std::cout << "Result sample (0,0): " << std::fixed << std::setprecision(6) << result[0][0] << std::endl;
}

int main() {
    run_benchmark();
    return 0;
}

File: rust/src/matrix.rs
use rand::Rng;
use std::time::Instant;

/// Matrix multiplication implementation
fn matrix_multiply(matrix_a: &Vec<Vec<f64>>, matrix_b: &Vec<Vec<f64>>) -> Vec<Vec<f64>> {
    let rows_a = matrix_a.len();
    let cols_a = matrix_a[0].len();
    let cols_b = matrix_b[0].len();

    // Initialize result matrix with zeros
    let mut result = vec![vec![0.0; cols_b]; rows_a];

    // Perform multiplication
    for i in 0..rows_a {
        for j in 0..cols_b {
            for k in 0..cols_a {
                result[i][j] += matrix_a[i][k] * matrix_b[k][j];
            }
        }
    }

    result
}

/// Run matrix multiplication benchmark with 100x100 matrices
pub fn run_benchmark() {
    let size = 100;
    let mut rng = rand::thread_rng();

    // Generate two random 100x100 matrices
    let matrix_a: Vec<Vec<f64>> = (0..size)
        .map(|_| (0..size).map(|_| rng.gen::<f64>()).collect())
        .collect();
    let matrix_b: Vec<Vec<f64>> = (0..size)
        .map(|_| (0..size).map(|_| rng.gen::<f64>()).collect())
        .collect();

    // Measure multiplication time
    let start_time = Instant::now();
    let result = matrix_multiply(&matrix_a, &matrix_b);
    let execution_time = start_time.elapsed().as_millis();

    println!("Test: Matrix Multiplication");
    println!("Matrix size: {}x{}", size, size);
    println!("Execution time: {} ms", execution_time);
    println!("Result sample (0,0): {:.6}", result[0][0]);
}
